### PR TITLE
Fix nonce issue when submitting multiple processes

### PR DESCRIPTION
**Pull request by mattdean-digicatapult**
# Pull Request

## Checklist
- [x] Have you read Digital Catapult's [Code of Conduct](https://github.com/digicatapult/.github/blob/main/CODE_OF_CONDUCT.md)?
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] New and existing unit tests pass locally with my changes.

## PR Type

Please delete options that are irrelevant.

- [x] Bug Fix
- [ ] Chore
- [ ] Feature
- [ ] Documentation Update
- [ ] Code style update (formatting, local variables)
- [ ] Breaking Change (fix or feature that would cause existing functionality to change)

## Linked tickets

https://digicatapult.atlassian.net/browse/SQNC-49

## High level description

Fixes issue where trying to load multiple process flows leads to nonce errors

## Detailed description

When trying to load multiple flows from matchmaker we've been encountering errors like 

```
RPC-CORE: submitAndWatchExtrinsic(extrinsic: Extrinsic): ExtrinsicStatus:: 1014: Priority is too low: (0 vs 0): The transaction has too low priority to replace another transaction already in the pool.
```

This is caused by incorrect nonce handling and was introduced by polkadot 12 as per https://github.com/digicatapult/sqnc-matchmaker-api/pull/409. This fixes the issue in the same way.

I've additionally moved our test suite to `testcontainers` so I could write a test for this issue.

Lastly I've noticed that polkadot was not cleaning up connections properly so I've introduced resource cleanup using an async dispose pattern. I suspect this may be what we are hitting in https://digicatapult.atlassian.net/browse/SQNC-24

## Describe alternatives you've considered

None

## Operational impact

None

## Additional context

None

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -200,14 +200,14 @@ Unit tests can be run without docker using:
...
-To run the integration test suite first bring up the test dependency services using docker
+Running the integration test suite requires docker to be installed. Tests can then be executed with:
 
 ```shell
-docker compose -f ./docker-compose-test.yaml up -d
+npm run test
 ```
 
-And then to run the tests
+If you with to see output from the `sqnc-node` container brought up with `testcontainers` run:
```
Comment (Ellenn-A) [inline README.md:209]: if you want?

File: src/lib/process/index.ts [hunk 2]
```diff
@@ -243,7 +243,7 @@ export const disableProcess = async (
 ): Promise<Process.ProcessResponse> => {
   const processId = utf8ToHex(name)
 
-  const polkadot: Polkadot.Polkadot = await createNodeApi(options)
+  await using polkadot = await createNodeApi(options)
```
Comment (Ellenn-A) [inline src/lib/process/index.ts:246]: so the `using` uses this (from he file below): 
```
[Symbol.asyncDispose]: async () => {
      await api.disconnect()
    },
```
to dispose of the polkadot instance/disconnect it once this block finishes? (that's the explanation I found)

Comment (mattdean-digicatapult) [reply src/lib/process/index.ts:246]: the syntax is terrifying but the behaviour is awesome

File: tests/integration/without-manual-seal.test.ts [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+import { type StartedTestContainer } from 'testcontainers'
+
+import { loadProcesses } from '../../src/lib/process/index.js'
+import { multiple } from '../fixtures/programs.js'
+
+import { getVersionHelper } from '../helpers/substrateHelper.js'
+
+import { withSubstrateNode } from '../helpers/containers.js'
+
+describe('Transactions without manual seal', function () {
```
Comment (Ellenn-A) [inline tests/integration/without-manual-seal.test.ts:11]: what is a manual seal out of interest

File: src/lib/process/api.ts [hunk 4]
```diff
@@ -12,6 +12,11 @@ const isProgramValid = (program: Process.Program, out = { ops: 0, restrictions:
   return out.ops === out.restrictions
 }
 
+let lastSubmittedNonce: number = -1
+export function resetLastSubmittedNonce() {
```
Comment (Ellenn-A) [inline src/lib/process/api.ts:16]: is this only for the tests?

File: tests/integration/without-manual-seal.test.ts (hunk 3 above)
Comment (mattdean-digicatapult) [reply tests/integration/without-manual-seal.test.ts:11]: so the node generally produces blocks every 6s. These are then finalised usually after another 6s-18s. Only then is the transaction final. Manual seal allows us to call an API on the node and manually finalise the blocks. This makes the tests much much faster but was masking this issue. Hence I've added one test not using manual seal to catch further issues down the road

File: src/lib/process/api.ts (hunk 4 above)
Comment (mattdean-digicatapult) [reply src/lib/process/api.ts:16]: for now yes. I can forsee conditions where we'd need to do this elsewehere

**Review by Ellenn-A — COMMENTED** [review]
another question (just to re-iterate I understand) so the reason we didn't come across this issue: `... that polkadot was not cleaning up connections properly...` is because in ci between each tests the polkadot connection would reset itself/make a new connection? 

So the same was happening locally? The tests would have just passed cuz each time polkadot connection would get created before/during a test and then disposed of after test?

**Comment by mattdean-digicatapult** [comment]
They were passing because the process was always exiting naturally before we exhausted the maximum number of node connections

**Review by Ellenn-A — APPROVED** [review]
lgtm

**Pull request merged**